x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/async/AsyncSearchIndexServiceTests.java

```diff
@@ -264,4 +270,38 @@ public CircuitBreakerStats stats(String name) {
             }
         }
     }
+
+    public void testMaxAsyncSearchResponseSize() throws Exception {
+        // setting very small limit for the max size of async search response
+        int limit = randomIntBetween(1, 125);
+        ClusterUpdateSettingsRequest updateSettingsRequest = new ClusterUpdateSettingsRequest();
+        updateSettingsRequest.transientSettings(Settings.builder().put("search.max_async_search_response_size", limit + "b"));
+        assertAcked(client().admin().cluster().updateSettings(updateSettingsRequest).actionGet());
+
+        AsyncExecutionId executionId = new AsyncExecutionId(Long.toString(randomNonNegativeLong()),
+            new TaskId(randomAlphaOfLength(10), randomNonNegativeLong()));
+
+        PlainActionFuture<IndexResponse> createFuture = new PlainActionFuture<>();
+        TestAsyncResponse initialResponse = new TestAsyncResponse(randomAlphaOfLength(130), randomLong());
+        indexService.createResponse(executionId.getDocId(), Map.of(), initialResponse, createFuture);
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, createFuture::actionGet);
+        assertEquals("Can't store an async search response larger than ["+ limit + "] bytes. " +
+                "This limit can be set by changing the [" + MAX_ASYNC_SEARCH_RESPONSE_SIZE_SETTING.getKey() + "] setting.",
+            e.getMessage());
+        assertEquals(0, e.getSuppressed().length); // no other suppressed exceptions
+
+        PlainActionFuture<UpdateResponse> updateFuture = new PlainActionFuture<>();
+        TestAsyncResponse updateResponse = new TestAsyncResponse(randomAlphaOfLength(130), randomLong());
+        indexService.updateResponse(executionId.getDocId(), Map.of(), updateResponse, updateFuture);
+        e = expectThrows(IllegalArgumentException.class, createFuture::actionGet);
+        assertEquals("Can't store an async search response larger than ["+ limit + "] bytes. " +
+                "This limit can be set by changing the [" + MAX_ASYNC_SEARCH_RESPONSE_SIZE_SETTING.getKey() + "] setting.",
+            e.getMessage());
+        assertEquals(0, e.getSuppressed().length); // no other suppressed exceptions
+
```

Comment:
Can we have a test verifying that we have recorded the failure in the async search index?

Reply:
Good point,  addressed in d03015c